crates/flux-syntax/src/parser/mod.rs

```diff
@@ -94,29 +97,232 @@ pub(crate) fn parse_flux_items(cx: &mut ParseCtxt) -> ParseResult<Vec<Item>> {
 ///              | ⟨sort_decl⟩
 ///              | ⟨primop_prop⟩
 /// ```
-fn parse_flux_item(cx: &mut ParseCtxt) -> ParseResult<Item> {
+fn parse_flux_item(cx: &mut ParseCtxt) -> ParseResult<FluxItem> {
     let mut lookahead = cx.lookahead1();
     if lookahead.peek(token::Pound) || lookahead.peek(kw::Fn) {
-        parse_reft_func(cx).map(Item::FuncDef)
+        parse_reft_func(cx).map(FluxItem::FuncDef)
     } else if lookahead.peek(kw::Local) || lookahead.peek(kw::Qualifier) {
-        parse_qualifier(cx).map(Item::Qualifier)
+        parse_qualifier(cx).map(FluxItem::Qualifier)
     } else if lookahead.peek(kw::Opaque) {
-        parse_sort_decl(cx).map(Item::SortDecl)
+        parse_sort_decl(cx).map(FluxItem::SortDecl)
     } else if lookahead.peek(kw::Property) {
-        parse_primop_property(cx).map(Item::PrimOpProp)
+        parse_primop_property(cx).map(FluxItem::PrimOpProp)
     } else {
         Err(lookahead.into_error())
     }
 }
 
-fn parse_hide_attr(cx: &mut ParseCtxt) -> ParseResult<bool> {
-    if !cx.advance_if(token::Pound) {
-        return Ok(false);
+///```text
+/// ⟨specs⟩ ::= ⟨specs⟩*
+/// ```
+pub(crate) fn parse_detached_specs(cx: &mut ParseCtxt) -> ParseResult<surface::DetachedSpecs> {
+    let items = until(cx, token::Eof, parse_detached_item)?;
+    Ok(surface::DetachedSpecs { items })
+}
+
+///```text
+/// ⟨specs⟩ ::= ⟨fn-spec⟩
+///           | ⟨struct-spec⟩
+///           | ⟨enum-spec⟩
+///           | mod  <PATH> { ⟨specs⟩ }
+///           | impl <PATH> { ⟨impl-spec⟩ }
+/// ```
+pub(crate) fn parse_detached_item(cx: &mut ParseCtxt) -> ParseResult<Item> {
+    let mut lookahead = cx.lookahead1();
+    if lookahead.peek(kw::Fn) || lookahead.peek(token::Pound) {
+        let item = parse_detached_fn_sig(cx)?;
+        let ident = item.ident;
+        let kind = ItemKind::FnSig(item);
+        Ok(Item { ident, kind })
+    } else if lookahead.peek(kw::Mod) {
+        parse_detached_mod(cx)
+    } else if lookahead.peek(kw::Struct) {
+        parse_detached_struct(cx)
+    } else if lookahead.peek(kw::Enum) {
+        parse_detached_enum(cx)
+    } else if lookahead.peek(kw::Impl) {
+        parse_detached_impl(cx)
+    } else if lookahead.peek(kw::Trait) {
+        parse_detached_trait(cx)
+    } else {
+        Err(lookahead.into_error())
+    }
+}
+
+///```text
+/// ⟨field⟩ ::= Ident : ⟨Ty⟩
+/// ```
+fn parse_detached_field(cx: &mut ParseCtxt) -> ParseResult<(Ident, Ty)> {
+    let ident = parse_ident(cx)?;
+    cx.expect(token::Colon)?;
+    let ty = parse_type(cx)?;
+    Ok((ident, ty))
+}
+
+///```text
+/// ⟨refine_info⟩ ::= [(Ident : ⟨sort⟩)*]?
+///                   invariant(<expr>)?
+/// ```
+fn parse_detached_refine_info(
```

Comment:
I am now leaning towards parsing `refined_by` and `invariant` as attributes. I don't think you gain much readability from the "inline" syntax, and probably the less we deviate from the attached syntax, the better.

Reply:
done.